src/base/bittorrent/session.cpp

```diff
@@ -371,6 +371,9 @@ Session::Session(QObject *parent)
     , m_useProxy(false)
     , m_recentErroredTorrentsTimer(new QTimer(this))
 {
+    if (useRandomPort() || (port() < 0))
+        m_port = Utils::Random::rand(1024, 65535);
```

Comment:
It would probably be better to assign a port in the dynamic port range, to avoid conflicts with other well-known assignments.
https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml

```suggestion
        m_port = Utils::Random::rand(49152, 65535);
```

Reply:
>to avoid conflicts with other well-known assignments.

I think the probability of conflict would be low, also after seeing this [SO answer](https://stackoverflow.com/a/38141340) I believe qbt should avoid getting ports *solely* from the range 49152 - 65535.
I'll leave the discussion (and code change) for another PR.

Reply:
After reading the SO page you linked and thinking about it a bit more, I realize now that the proper fix would be to set qBittorrent to choose a random port in the IANA dynamic range _at each startup by default._

This way, the average user will never run into problems. Those that need qBittorrent to always listen on a specific port can do as they see fit.

What do you think about this?

Reply:
>I realize now that the proper fix would be to set qBittorrent to choose a random port in the IANA dynamic range at each startup by default.

I'm not sure or confident that should be on by default. 

>for clarity, what I mean is to let libtorrent request an unused port from the OS

If you mean binding to port `0`, then yes, that should have the least problems.
I just tried, one side effect is the TCP port and UDP port will be different most of the time, would that pose a problem? Do users expect the ports to be the same?

Reply:
> I'm not sure nor confident that should be on by default.
> If you mean binding to port `0`, then yes, that should have the least problems.

Yes, that is what I mean, so why are not confident that it should be on by default, if, like you say also, it should have the least problems?

> I just tried, one side effect is the TCP port number and UDP port number will be different most of the time, would that pose a problem? Do users expect the ports to be the same?

The average user does not care. Those who have the need to set a specific port can already do so in the client's settings.

In fact, funnily enough, you can specify a specific persistent port number for TCP and UDP to listen on, but I don't think there is a way for a user to specify different persistent ports for TCP and UDP in qBittorrent.

Reply:
>In fact, funnily enough ...

This part worries me, why is it designed like this? Are there something hidden beneath it?

>Yes, that is what I mean, so why are you not confident that it should be on by default, if, like you say also, it should have the least problems?

The conscious of having the least problems is merely based on my speculation and the old functionality has been working reliably for ages, would tweaking it to act another way cause unnecessary problems? that I have no answer. 

Anyway I've already added a second commit to the PR, you might want to test it a bit, see if it is as good as before.

Reply:
I looked at the current code and some things look weird to me.

For example, doesn't the last line in this set of statements always evaluate to `settingsPack.set_int(libt::settings_pack::max_retry_port_bind, 0);`?

```cpp
const ushort port = this->port();
const std::pair<int, int> ports(port, port);
settingsPack.set_int(libt::settings_pack::max_retry_port_bind, ports.second - ports.first);
```

It seems as though the intention was to make a pair with `port` and something else.

---

Also, the `Session::port()` method reads to me as a "getter", but at the same time it does another thing (it generates a new random port), which I think can pottentially lead to bugs.

I see that you have changed the `Session::setPort(const int port)` method to no longer depend on it in this PR, but perhaps the `Session::port()` method should actually be corrected as well? The part about generating a random port can be refactored somewhere else.

Reply:
>I looked at the current code and some things look weird to me.

Must be history remnants, no idea whatsoever.

>but perhaps the Session::port() method should actually be corrected as well? 

I already did in the first commit, or do you mean something else?

>Anyway I've already added a second commit to the PR, you might want to test it a bit, see if it is as good as before.

To be precise, I meant I wish you (or anyone out there) could actually compile and run and see how it goes, I could test some linux ISO in the public but I don't use private trackers nor have experiences with them.

Reply:
> I already did in the first commit, or do you mean something else?

I mean this, line 2452 as of writing:

```cpp
int Session::port() const
{
    static int randomPort = Utils::Random::rand(1024, 65535);
    if (useRandomPort())
        return randomPort;
    return m_port;
}
```

> To be precise, I meant I wish you (or anyone out there) could actually compile and run and see how it goes, I could test some linux ISO in the public but I don't use private trackers nor have experiences with them.

Sure. Do I need to do anything special about libtorrent compilation? I am using Xubuntu 18.04.2 LTS.

Reply:
>I mean this, line 2452 as of writing:

What should I do about it? you just pasted the old code, isn't it? If you mean reverting it, then no, the first commit requires it to be that way...

>Do I need to do anything special about libtorrent compilation? 

I guess not.

Reply:
> you just pasted the old code, isn't it? 
> If you mean reverting it, then no, the first commit requires it to be that way...

yes, I think it should be changed, which actually would require changes to the code for this PR as well. Maybe you can leave that change for another PR then?

Reply:
Using a `static` var here is actually wrong conceptually, it prohibits `Session` to randomly select another port in case of closing the current session and creating another session instance.

>Maybe you can leave that change for another PR then?

OK, let's just ignore it for now.

Reply:
Ok, so I compiled your branch (and also compiled libtorrent from the latest commit of the RC_1_2 branch) and when I ran qbittorrent, the port still defaulted to 8999. I made sure to reset the settings beforehand (deleted the directories `.local/share/data/qBittorrent/` and `.config/qBittorrent/`).

Any ideas?

Reply:
@Chocobo1 
Ok, I finished testing. Everything seems to work fine. This will not affect users with existing installations, unless they clear the settings of course.

It still think it would be ideal for qBittorrent to choose a random port by default at startup. But that can be addressed in another PR.

Oh, and by the way, the current behaviour (even with your new code) of TCP vs UDP ports is the following:
- If the user selects a port manually (even if using the "Random" button in the UI), qBttorrent will listen on the same port for both TCP and UDP.
-if the user ticks the "Use a differrent port each startup", qBittorrent will listen on different ports for TCP and UDP (though there is a small chance that they coincide, unless whatever code path is being taken somewhere forbids them to be the same, which I doubt).

In all cases qBittorrent appears to behave properly.

Reply:
By the way, this probably does not matter, but my tests were conducted in a network with IPv4 only, and no UPnP/NAT-PMP. The failure of UPnP/NAT-PMP port mapping is due to my network, not to anything related to this specific branch of qBittorrent.
```
External IP: <redacted IPv4>
UPnP/NAT-PMP: Port mapping failure, message: could not map port using UPnP: no router found
```